src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAspect.cs

```diff
@@ -0,0 +1,378 @@
+#nullable enable
```

Comment:
Woah, this was obviously very difficult to implement, well done! 💯 🥇 
Nice work being able to box and unbox the different types.
I think rather than try and understand everything I'm going to accept this and go through understanding it further as I need to interact with it, as I already spent quite a lot of time on this PR.

Just one question here, this looks like it will cache an ActionResult that is a failure, that we might then not clear until we invalidate the cache key, whereas we might want to retry it if it's an intermittent failure. Is that correct?

Reply:
I don't think this should be the case. I'm guessing you probably know, but just in case - the `ActionResult` and `ActionResult<T>` types are really different kettles of fish and imply different things. Typically we would need to do `new ActionResult(something)` to get an `ActionResult<T>`, whereas `ActionResult` is typically created using derivatives such as `OkResult` or `NotFoundResult`.

`ActionResult` is like a 'void' type and should not have an internal value to return. Consequently, our unbox/boxing logic only applies for `ActionResult<T>` as this represents a boxed value T.

I think we prove that this works as expected in the `NoParams_ActionResult` and `NoParams_ActionResult_CacheMiss_NotFound` test cases in `CacheAspectTests`.

Of course, if you wanted to, you could try and cache a method like:

```csharp
[BlobCache(typeof(MyKey))]
public ActionResult DoSomething()
```

In this case, we would end up caching the `ActionResult` itself. A similar thing would happen if we used `Task` instead of `Task<T>` as well. However, it should be obvious why returning a void `ActionResult` or `Task` and expecting it to be cached doesn't really make sense - it would be like caching a `void` method.

We actually throw an error if the return type is void, but we currently don't do this for void `ActionResult` or `Task` because we can only assume the consumer actually wishes to serialize the `ActionResult` or `Task` objects into the cache for whatever strange reason.

Reply:
Thanks, I missed that but taken a closer look now. `CacheAspectTests` looks pretty exhaustive!